CRM/Case/ActionMapping.php

```diff
@@ -0,0 +1,138 @@
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ * Class CRM_Event_ActionMapping
+ *
+ * This defines the scheduled-reminder functionality for CiviCase entities.
+ */
+class CRM_Case_ActionMapping extends \Civi\ActionSchedule\Mapping {
+
+  /**
+   * The value for civicrm_action_schedule.mapping_id which identifies the
+   * "Case" mapping.
+   *
+   * Note: This value is chosen to match legacy DB IDs.
+   */
+  const CASE_MAPPING_ID = 99;
+
+  /**
+   * Register Case-related action mappings.
+   *
+   * @param \Civi\ActionSchedule\Event\MappingRegisterEvent $registrations
+   */
+  public static function onRegisterActionMappings(\Civi\ActionSchedule\Event\MappingRegisterEvent $registrations) {
+    $registrations->register(CRM_Case_ActionMapping::create(array(
+      'id' => CRM_Case_ActionMapping::CASE_MAPPING_ID,
+      'entity' => 'civicrm_case',
+      'entity_label' => ts('Case'),
+      'entity_value' => 'case_type',
+      'entity_value_label' => ts('Case Type'),
+      'entity_status' => 'case_status',
+      'entity_status_label' => ts('Case Status'),
+      'entity_date_start' => 'case_start_date',
+      'entity_date_end' => 'case_end_date',
+    )));
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function getRecipientListing($type) {
+    if ($type == 'case_roles') {
+      $result = civicrm_api3('RelationshipType', 'get', array(
+        'sequential' => 1,
+        'options' => array('limit' => 0, 'sort' => "label_b_a"),
+      ))['values'];
+      $cRoles = array();
+      foreach ($result as $each) {
+        $cRoles[$each['id']] = $each['label_b_a'];
+      }
+      return $cRoles;
+    }
+    return array();
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function getRecipientTypes() {
+    return array('case_roles' => 'Case Roles');
+  }
+
+  /**
+   * @inheritdoc
+   */
+  public function createQuery($schedule, $phase, $defaultParams) {
+    $selectedValues = (array) \CRM_Utils_Array::explodePadded($schedule->entity_value);
+    $selectedStatuses = (array) \CRM_Utils_Array::explodePadded($schedule->entity_status);
+
+    $query = \CRM_Utils_SQL_Select::from("civicrm_case c")->param($defaultParams);
+    $query->join('r', "INNER JOIN civicrm_relationship r ON c.id = r.case_id");
+    $query->join('cs', "INNER JOIN civicrm_contact ct ON r.contact_id_b = ct.id");
+    $query->join('ce', "INNER JOIN civicrm_email ce ON ce.contact_id = ct.id");
+
+    if (!empty($selectedValues)) {
+      $query->where("c.case_type_id IN (#caseId)")
+        ->param('caseId', $selectedValues);
+    }
+
+    if (!empty($selectedStatuses)) {
+      $query->where("c.status_id IN (#selectedStatuses)")
+        ->param('selectedStatuses', $selectedStatuses);
+    }
+
+    if (!empty($caseRoles)) {
```

Comment:
Can't seem to find where the `$caseRoles` parameter is set in this function?

Reply:
this was not required here and is handled later in the function. I have removed this now.